src/communication/v2/session.hpp

```diff
@@ -0,0 +1,331 @@
+            [this](const TCPSocket &socket) { return execution_active_ && socket.is_open(); }},
+        socket_);
+  }
+
+ private:
+  explicit Session(tcp::socket &&socket, TSessionData *data, ServerContext &server_context, tcp::endpoint endpoint,
+                   int inactivity_timeout_sec, std::string_view service_name)
+      : socket_(CreateWebSocket(std::move(socket), server_context)),
+        strand_{boost::asio::make_strand(GetExecutor())},
+        output_stream_([this](const uint8_t *data, size_t len, bool have_more) { return Write(data, len, have_more); }),
+        session_(data, endpoint, input_buffer_.read_end(), &output_stream_),
+        endpoint_{endpoint},
+        remote_endpoint_{GetRemoteEndpoint()},
+        service_name_{service_name},
+        timeout_seconds_(inactivity_timeout_sec),
+        timeout_timer_(GetExecutor()) {
+    std::visit(
+        utils::Overloaded{[](SSLSocket &socket) {
+                            socket.lowest_layer().set_option(boost::asio::ip::tcp::no_delay(true));  // enable PSH
+                            socket.lowest_layer().set_option(
+                                boost::asio::socket_base::keep_alive(true));  // enable SO_KEEPALIVE
+                            socket.lowest_layer().non_blocking(false);
+                          },
+                          [](TCPSocket &socket) {
+                            socket.set_option(boost::asio::ip::tcp::no_delay(true));        // enable PSH
+                            socket.set_option(boost::asio::socket_base::keep_alive(true));  // enable SO_KEEPALIVE
+                            socket.non_blocking(false);
+                          }},
+        socket_);
+    timeout_timer_.expires_at(boost::asio::steady_timer::time_point::max());
+    spdlog::info("Accepted a connection from {}:", service_name_, remote_endpoint_.address(), remote_endpoint_.port());
+  }
+
+  void DoWrite(const uint8_t *data, size_t len, bool /*have_more*/) {
+    if (!IsConnected()) {
+      return;
+    }
+    std::visit(utils::Overloaded{[shared_this = this->shared_from_this(), data, len](TCPSocket &socket) {
+                                   boost::asio::write(socket, boost::asio::buffer(data, len));
+                                 },
+                                 [shared_this = this->shared_from_this(), data, len](SSLSocket &socket) {
+                                   boost::asio::write(socket, boost::asio::buffer(data, len));
+                                 }},
+               socket_);
+  }
+
+  void DoRead() {
+    if (!IsConnected()) {
+      return;
+    }
+    timeout_timer_.expires_after(std::chrono::seconds(this->timeout_seconds_));
+    ExecuteForSocket([this](auto &&socket) {
+      auto buffer = input_buffer_.write_end()->Allocate();
+      socket.async_read_some(
+          boost::asio::buffer(buffer.data, buffer.len),
+          boost::asio::bind_executor(strand_, std::bind_front(&Session::OnRead, this->shared_from_this())));
+    });
```

Comment:
so funny thing the `boost::asio::async_read` does not work in our case and here is my reasoning behind this. In order to ensure that we have read everything we need to know how much is this everything therefore the `boost::asio::async_read` must have and indication of the size of the data it will read, there are three options:
- `transfer_all()` => which is a default one in which we need to receive the whole buffer in order to confirm that we have read anything, the issue is that bolt messages rarely fill the whole bugger, they are dynamic in size
- `transfer_at_least(size_t)` => to which we pass the minimum size of the buffer that signals that we have read something, this in my opinion ends up being same as `async_read_some`
- `transfer_exactly(size_t)` => to which we need to pass to exact size we want to read, which again is unknown
the